app/src/main/java/edu/uc/kovaciad/slicetracker/dao/SliceDatabase.kt

```diff
@@ -9,7 +10,65 @@ import edu.uc.kovaciad.slicetracker.dto.Printer
-}
+
+    /**
+     * Define a companion object, this allows us to add functions on the SliceDatabase class.
+     */
+    companion object {
+        /**
+         * INSTANCE will keep a reference to any database returned via getInstance.
+         *
+         * This will help us avoid repeatedly initializing the database, which is expensive.
+         *
+         *  The value of a volatile variable will never be cached, and all writes and
+         *  reads will be done to and from the main memory. It means that changes made by one
+         *  thread to shared data are visible to other threads.
+         */
+        @Volatile
+        private var INSTANCE: SliceDatabase? = null
+        /**
+         * Helper function to get the database.
+         * If a database has already been retrieved, the previous database will be returned.
+         * Otherwise, create a new database.
+         * This function is threadsafe, and callers should cache the result for multiple database
+         * calls to avoid overhead.
+         * This is an example of a simple Singleton pattern that takes another Singleton as an
+         * argument in Kotlin.
+         * To learn more about Singleton read the wikipedia article:
+         * https://en.wikipedia.org/wiki/Singleton_pattern
+         *
+         * @param context The application context Singleton, used to get access to the filesystem.
+         */
+        fun getInstance(context: Context): SliceDatabase {
+            // Multiple threads can ask for the database at the same time, ensure we only initialize
+            // it once by using synchronized. Only one thread may enter a synchronized block at a
+            // time.
+            synchronized(this) {
+                // Copy the current value of INSTANCE to a local variable so Kotlin can smart cast.
+                // Smart cast is only available to local variables.
+                var instance = INSTANCE
+                // If instance is `null` make a new database instance.
+                if (instance == null) {
+                    instance = Room.databaseBuilder(
+                        context.applicationContext,
+                        SliceDatabase::class.java,
+                        " "
+                    )
+                        // Wipes and rebuilds instead of migrating if no Migration object.
+                        // Migration is not part of this lesson. You can learn more about
+                        // migration with Room in this blog post:
+                        // https://medium.com/androiddevelopers/understanding-migrations-with-room-f01e04b07929
+                        .fallbackToDestructiveMigration()
+                        .build()
+                    // Assign INSTANCE to the newly created database.
+                    INSTANCE = instance
+                }
+                // Return instance; smart cast to be non-null.
+                return instance
+            }
+        }
+    }
+}
```

Comment:
I defined a companion object, this allows you to add functions on the SliceDatabase class. INSTANCE will keep a reference to any database returned via getInstance. This will help you avoid repeatedly initializing the database, which is expensive. The value of a volatile variable will never be cached, and all writes and reads will be done to and from the main memory. It means that changes made by one thread to shared data are visible to other threads.
Check out: https://kotlinlang.org/api/latest/jvm/stdlib/kotlin.jvm/-volatile/
https://www.baeldung.com/kotlin/volatile-properties
https://kotlinlang.org/api/latest/jvm/stdlib/kotlin.reflect/-k-parameter/-kind/-i-n-s-t-a-n-c-e.html